recovery/cmd/backup_size.go

```diff
@@ -0,0 +1,189 @@
+package cmd
+
+import (
+	"errors"
+	"io/fs"
+	"os"
+	"path/filepath"
+
+	disk "github.com/shirou/gopsutil/v3/disk"
+	log "github.com/sirupsen/logrus"
+)
+
+var (
+	clusterPath         = "/var/lib/etcd/member/snap/db"
+	staticPodsPath      = "/etc/kubernetes/static-pod-resources/"
+	usrLocalPath        = "/usr/local"
+	kubeletPath         = "/var/lib/kubelet"
+	etcPath             = "/etc"
+	binPath             = "/etc/kubernetes/static-pod-resources/bin/"
+	backupSizeSafetyNet = 10.00
+	units               = []string{"B", "KiB", "MiB", "GiB", "TiB", "EiB", "ZiB"}
+)
+
+type resource struct {
+	dirPath string
+}
+
+type resourceList struct {
+	resources *[]resource
+}
+
+// compareBackupToDisk verifies disk space required against available
+// disk space and returns boolean value
+func compareBackupToDisk() (bool, error) {
+
+	estimated, err := estimateDiskSize()
+	if err != nil {
+		log.Errorf("Couldn't calculate estimated disk space")
+		return false, err
+	}
+
+	freeDisk, err := diskPartitionSize()
+	if err != nil {
+		log.Errorf("Couldn't calculate free disk space")
+		return false, err
+	}
+
+	estimated, ebytes := sizeConversion(estimated)
+	freeDisk, fbytes := sizeConversion(freeDisk)
+
+	log.Infof("Available disk space : %.2f %s; Estimated disk space required for backup: %.2f %s \n", freeDisk, fbytes, estimated, ebytes)
+
+	// find the index of estimated and freedisk space in units[]
+	var ebytesID, fbytesID int
+	for i := range units {
+		if units[i] == ebytes {
+			ebytesID = i
+		}
+		if units[i] == fbytes {
+			fbytesID = i
+		}
+	}
+
+	if ebytesID <= fbytesID {
+		// safetynet of 10GiB is added
+		if freeDisk > backupSizeSafetyNet+estimated {
+			return true, nil
+		}
+	}
+	return false, nil
+}
+
+// estimateDiskSize calculate the required backup size
+// returns: disk size(int), error
+func estimateDiskSize() (float64, error) {
+
+	DirList := resourceList{
+		&[]resource{{staticPodsPath},
+			{clusterPath},
+			{usrLocalPath},
+			{kubeletPath},
+			{etcPath},
+		},
+	}
+
+	estDirMap := map[string]float64{}
+	var total float64
+
+	for _, v := range *DirList.resources {
+		_, err := os.Lstat(v.dirPath)
+		if err != nil {
+			return total, err
+		}
+
+		switch v.dirPath {
+		case staticPodsPath:
+			binPathSize := dirSize(binPath)
+			estDirMap[v.dirPath] = dirSize(v.dirPath) - binPathSize
+
+		case clusterPath:
+			estDirMap[v.dirPath] = dirSize(v.dirPath) + estDirMap[staticPodsPath]
+
+		default:
+			estDirMap[v.dirPath] = dirSize(v.dirPath)
+		}
+
+		total += estDirMap[v.dirPath]
+
+	}
+	return total, nil
+}
+
+// diskPartitionSize calculate current disk space
+// returns:     disk size(int), error
+func diskPartitionSize() (float64, error) {
+
+	var diskPartition []disk.PartitionStat
+
+	diskPartition, err := disk.Partitions(false)
+	if err != nil {
+		log.Println(err)
+	}
+
+	var totalDiskSize, freeDiskSpace float64
+	for _, v := range diskPartition {
```

Comment:
Are you totalling all partitions in this loop? I'd have expected we're just looking at /var/recovery

Reply:
If the extra partition is not created and /var/recovery is not a separate partition then what should we do?

Reply:
We'd just care about the free space available under /var/recovery, whether it's a separate partition or not

Reply:
changed as suggested with commit effab4c9023ada38a70df52005e59d4bda2a8714